### DISCUSSION TITLE
How to get Tooltip and MenuButton playing nice together

### DISCUSSION THREAD

**Discussion by JakeGinnivan** [chakra-ui/chakra-ui#3193] [Q&A] [answered]
Hey all,

Been trying to get a Tooltip working with a menu button but not had a lot of luck. My guess is something around refs but I just can't seem to get the right combination, even with creating a mergeRef component to try and merge the two refs.

Any pointers would be great

**Comment by dodas** [comment, marked as answer]
Hey, it seems to work fine in the docs, see the gif. I am misunderstanding something?
![menu-tooltip](https://user-images.githubusercontent.com/14360171/108513427-e2d89a00-72c2-11eb-982a-c26739a7ae5a.gif)

**Reply by rahmadPoldit** [reply]
I'm having the same issue.  My issue deals with the picture you provided @dodas .  When you click on the menu button, it opens the dropdown and the tooltip disappears.  However when you click off the menu list and the list disappears, the tooltip still stays and you have to click to hide it.  How would we go about hiding the tooltip when the menu contents disappear?

**Reply by RyanWarner** [reply]
@rahmadPoldit I have the same issue. Did you find a solution?

**Reply by SugarDarius** [reply]
@RyanWarner A good solution is to control the `Tooltip` and the `Menu` directly. It's a bit of more code but it works very well.
Take a look at this simple example: 

```tsx
function MainMenu() {
    const { isOpen, onOpen, onClose } = useDisclosure()

    const [
      isHovering,
      setDelayedIsHoveringState,
      setIsHoveringState,
      cancelHoveringDelay,
    ] = useDelayedState<boolean>(false)
    
    const handleOnOpen = useCallback((): void => {
      onOpen()

      setIsHoveringState(false)
      cancelHoveringDelay()
    }, [onOpen, cancelHoveringDelay, setIsHoveringState])

   const onMouseEnter = useCallback((): void => {
     if (!isOpen) {
        setDelayedIsHoveringState(true)
      }
    }, [isOpen, setDelayedIsHoveringState])

    const onMouseLeave = useCallback(() => {
      setIsHoveringState(false)
      cancelHoveringDelay()
    }, [setIsHoveringState, cancelHoveringDelay])
    
    return (
        <Menu isOpen={isOpen} onOpen={handleOnOpen} onClose={onClose}>
            <Tooltip isOpen={isHovering}>
                <MenuButton onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />
            </Tooltip>
        </Menu>
    )
}
```

Hope it helps 🙏🏻

**Reply by lukasbach** [reply]
I think the tooltip showing after the menu was closed is intended, because after closing the menu, the button is focused and the tooltip shows for focused items by default.

**Comment by kasperadk** [comment]
Had the same problem, solved it this way:
```javascript

const [tooltipOpen, setTooltipOpen] = useState(false)

<Tooltip label="Open favorites" isOpen={tooltipOpen}>
  <MenuButton
    onClick={() => setTooltipOpen(false)}
    onMouseEnter={() => setTooltipOpen(true)}
    onMouseLeave={() => setTooltipOpen(false)}
    as={IconButton}
    aria-label="Open help menu"
    _hover={{ opacity: 0.75 }}
    variant="ghost"
    icon={
      <Icon
        as={FontAwesomeIcon}
        icon={faHeart}
        size="xl"
        cursor="pointer"
      />
    }
  />
</Tooltip>;
```

**Reply by AndreLuizGarcia** [reply]
Also, it's nice to also remember to add the support for Keyboard events

```js 
<Tooltip label="Open favorites" isOpen={tooltipOpen}>
  <MenuButton
    onClick={() => setTooltipOpen(false)}
    onMouseEnter={() => setTooltipOpen(true)}
    onMouseLeave={() => setTooltipOpen(false)}
    onKeyDown={() => setTooltipOpen(false)} // <-- this 
    onKeyUp={() => setTooltipOpen(true)}    // <-- this 
    as={IconButton}
    aria-label="Open help menu"
    _hover={{ opacity: 0.75 }}
    variant="ghost"
    icon={
      <Icon
        as={FontAwesomeIcon}
        icon={faHeart}
        size="xl"
        cursor="pointer"
      />
    }
  />
</Tooltip>;
